staging/src/k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing/queueset/queueset.go

```diff
@@ -0,0 +1,617 @@
+						req.Queue.Requests = append(req.Queue.Requests[:i],
+							req.Queue.Requests[i+1:]...)
+						break
+					}
+				}
+				// At this point, if the qs is quiescing,
+				// has zero requests executing, and has zero requests enqueued
+				// then a call to the EmptyHandler should be forked.
+				qs.maybeForkEmptyHandlerLocked()
+			} else {
+				klog.V(5).Infof("QS(%s): request %#+v %#+v canceled shortly after dispatch", qs.config.Name, descr1, descr2)
+			}
+		}
+		return decisionStr
+	}()
+	switch decision {
+	case DecisionTryAnother:
+		return true, false, func() {}
+	case DecisionReject:
+		return false, false, func() {}
+	case DecisionCancel:
+		return false, false, func() {}
+	default:
+		if decision != DecisionExecute {
+			klog.Errorf("Impossible decision %q", decision)
+		}
+		return false, true, func() {
+			qs.finishRequestAndDispatchAsMuchAsPossible(req)
+		}
+	}
+}
+
+// lockAndSyncTime acquires the lock and updates the virtual time.
+// Doing them together avoids the mistake of modify some queue state
+// before calling syncTimeLocked.
+func (qs *queueSet) lockAndSyncTime() {
+	qs.lock.Lock()
+	qs.syncTimeLocked()
+}
+
+// syncTimeLocked updates the virtual time based on the assumption
+// that the current state of the queues has been in effect since
+// `qs.lastRealTime`.  Thus, it should be invoked after acquiring the
+// lock and before modifying the state of any queue.
+func (qs *queueSet) syncTimeLocked() {
+	realNow := qs.clock.Now()
+	timesincelast := realNow.Sub(qs.lastRealTime).Seconds()
+	qs.lastRealTime = realNow
+	qs.virtualTime += timesincelast * qs.getVirtualTimeRatio()
+}
+
+// getVirtualTimeRatio calculates the rate at which virtual time has
+// been advancing, according to the logic in `doc.go`.
+func (qs *queueSet) getVirtualTimeRatio() float64 {
+	activeQueues := 0
+	reqs := 0
+	for _, queue := range qs.queues {
+		reqs += queue.RequestsExecuting
+		if len(queue.Requests) > 0 || queue.RequestsExecuting > 0 {
+			activeQueues++
+		}
+	}
+	if activeQueues == 0 {
+		return 0
+	}
+	return math.Min(float64(reqs), float64(qs.config.ConcurrencyLimit)) / float64(activeQueues)
+}
+
+// timeoutOldRequestsAndRejectOrEnqueueLocked encapsulates the logic required
+// to validate and enqueue a request for the queueSet/QueueSet:
+// 1) Start with shuffle sharding, to pick a queue.
+// 2) Reject old requests that have been waiting too long
+// 3) Reject current request if there is not enough concurrency shares and
+// we are at max queue length
+// 4) If not rejected, create a request and enqueue
+// returns the enqueud request on a successful enqueue
+// returns nil in the case that there is no available concurrency or
+// the queuelengthlimit has been reached
+func (qs *queueSet) timeoutOldRequestsAndRejectOrEnqueueLocked(hashValue uint64, descr1, descr2 interface{}) *request {
+	//	Start with the shuffle sharding, to pick a queue.
+	queueIdx := qs.chooseQueueIndexLocked(hashValue, descr1, descr2)
```

Comment:
Is it possible that qs.queues is empty ?
In that case -1 would be returned, leading to access error on the next line.

Reply:
That is not yet possible.  But when #84771 lands we will need some revisions here, to simply and clearly avoid queuing for a priority level that does limiting but not queuing.